server/src/postgres/simple_resource.rs

```diff
@@ -0,0 +1,404 @@
+    /// The data will be validated first, returning all possible errors at once.
+    /// If no validation errors are found, a new row will be inserted in the database and the new UUID will be returned.
+    async fn create<'a, T>(data: &T) -> Result<(Option<Uuid>, ValidationResult), ArrErr>
+    where
+        T: GrpcDataObjectType,
+    {
+        let (psql_data, validation_result) = Self::validate(data)?;
+
+        if !validation_result.success {
+            return Ok((None, validation_result));
+        }
+
+        let definition = Self::get_definition();
+        let id_col = Self::try_get_id_field()?;
+        let mut params: Vec<&PsqlField> = vec![];
+        let mut inserts = vec![];
+        let mut fields = vec![];
+        let mut index = 1;
+
+        for key in definition.fields.keys() {
+            match psql_data.get(&*key.to_string()) {
+                Some(value) => {
+                    let val: &PsqlField = <&Box<PsqlFieldSend>>::clone(&value).as_ref();
+                    fields.push(key.to_string());
+                    inserts.push(format!("${}", index));
+                    params.push(val);
+                    index += 1;
+                }
+                None => {
+                    psql_debug!(
+                        "Skipping insert [{}] for [{}] with data [{:?}]",
+                        key,
+                        definition.psql_table,
+                        data
+                    );
+                }
+            }
+        }
+        let insert_sql = &format!(
+            "INSERT INTO {} ({}) VALUES ({}) RETURNING {}",
+            definition.psql_table,
+            fields.join(", "),
+            inserts.join(", "),
+            id_col
+        );
+        psql_debug!("{}", insert_sql);
+
+        psql_info!("Inserting new entry for table [{}].", definition.psql_table);
+        let client = get_psql_pool().get().await.unwrap();
+        let row = client.query_one(insert_sql, &params[..]).await?;
+
+        Ok((Some(row.get(&*id_col)), validation_result))
+    }
+
+    /// Validates the given data against the resource definition.
+    /// Includes mandatory checks and type checks.
+    fn validate<T>(data: &T) -> Result<(PsqlData, ValidationResult), ArrErr>
+    where
+        T: GrpcDataObjectType,
+    {
+        let definition = Self::get_definition();
+
+        let mut converted: PsqlData = PsqlData::new();
+        let mut success = true;
+        let mut errors: Vec<ValidationError> = vec![];
+
+        // Only validate fields that are defined in self.definition.
+        // All other fields will be ignored (they will not be stored in the database either).
+        for (key, field) in definition.fields {
+            if field.is_internal() {
+                // internal field, skip for validation
+                continue;
+            }
+
+            let field_value = data.get_field_value(&key)?;
+            let val_to_validate = if field.is_mandatory() {
```

Comment:
this could be condensed to reduce nesting
```rust
let val_to_validate = match field_value {
    GrpcField::Option(option) => {
        let option: Option<GrpcField> = option.into();
        match option {
            Some(val) => val,
            None => {
                if field.is_mandatory() {
                    // Panic here, as this indicates a mismatch between the GRPC and the database definition for the resource.
                    panic!("Got 'GrpcField::Option' for [{}] [{:?}] while this field is not marked as optional in the definition.", key, field);
                }
                continue;
            }
        }
    }
    _ => {
        if !field.is_mandatory() {
            panic!("Expected 'GrpcField::Option' for [{}] [{:?}] since this field is marked as optional in the definition.", key, field);
        }
        field_value
    }
}
```

Reply:
Done!
https://github.com/Arrow-air/svc-storage/blob/6883178b0d80010b58ae192fab6479d96484803c/server/src/postgres/simple_resource.rs#L136